src/System.Private.CoreLib/src/System/Threading/FirstLevelSpinWaiter.cs

```diff
@@ -0,0 +1,99 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime;
+using Internal.Runtime.Augments;
+
+namespace System.Threading
+{
+    /// <summary>
+    /// A lightweight spin-waiter intended to be used as the first-level wait for a condition before the user forces the thread
+    /// into a wait state, and where the condition to be checked in each iteration is relatively cheap, like just an interlocked
+    /// operation.
+    /// 
+    /// Used by the wait subsystem on Unix, so this class cannot have any dependencies on the wait subsystem.
+    /// </summary>
+    internal struct FirstLevelSpinWaiter
+    {
+        // TODO: Tune these values
+        private const int SpinCount = 8;
+        private const int SpinYieldThreshold = 4;
+        private const int SpinSleep0Threshold = 6;
+
+        private static int s_processorCount;
+
+        private int _spinningThreadCount;
+
+        public void Initialize()
+        {
+            if (s_processorCount == 0)
+            {
+                s_processorCount = Environment.ProcessorCount;
+            }
+        }
+
+        public bool SpinWaitForCondition(Func<bool> condition)
+        {
+            Debug.Assert(condition != null);
+            Debug.Assert(s_processorCount > 0);
+
+            int processorCount = s_processorCount;
+            int spinningThreadCount = Interlocked.Increment(ref _spinningThreadCount);
+            try
+            {
+                // Limit the maximum spinning thread count to the processor count to prevent unnecessary context switching caused by
+                // an excessive number of threads spin waiting, perhaps even slowing down the thread holding the resource being
+                // waited upon
+                if (spinningThreadCount <= processorCount)
+                {
+                    // For uniprocessor systems, start at the yield threshold since the pause instructions used for waiting prior to
+                    // that threshold would not help other threads make progress
+                    for (int spinIndex = processorCount > 1 ? 0 : SpinYieldThreshold; spinIndex < SpinCount; ++spinIndex)
+                    {
+                        // The caller should check the condition in a fast path before calling this method, so wait first
+                        Wait(spinIndex);
+
+                        if (condition())
+                        {
+                            return true;
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                Interlocked.Decrement(ref _spinningThreadCount);
```

Comment:
It's new, I wanted to start with a light amount of spinning and go from there. The intention was to use this in any place where some spinning is done before doing a proper wait, especially in potentially high-contention cases like the process-wide lock used in the wait subsystem.